Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -15,11 +16,847 @@
+            UNIMPLEMENTED_WITH_MSG("Destination as Data Consumer is not handled right now");
+        }
+
+        RETURN_NULL_IF_FAILED(_idEncoder->Initialize(_idStream.Get(), WICBitmapEncoderNoCache));
+
+        if (_type == typeGIF) {
+            RETURN_NULL_IF_FAILED(_idEncoder->GetMetadataQueryWriter(&_idGifEncoderMetadataQueryWriter));
+        }
+    }
+
+    return self;
+}
+
+@end
+
+// Helper function for setting the property write values when formatting image metadata.
+// Image metadata stores decimal numbers as an integer divided by another integer.
+// The HighPart is a number representing what to divide the LowPart by to get the actual value.
+template<typename T>
+void setHighLowParts(T* out, double value) {
+    // Check to see if the value has a decimal component. If not, just divide by 1.
+    if (value == (int)value) {
+        (*out).LowPart = (int)value;
+        (*out).HighPart = 1;
+    } else {
+        (*out).LowPart = (int)(value * 100);
+        (*out).HighPart = 100;
+    }
+}
+
+// Helper function to write properties to image frames
+void writePropertyToFrame(PROPVARIANT* propertyToWrite, LPCWSTR path, IWICMetadataQueryWriter* propertyWriter) {
+    HRESULT status = propertyWriter->SetMetadataByName(path, propertyToWrite);
+    if (!SUCCEEDED(status)) {
+        NSString* pathNSString = [[NSString alloc] initWithBytes:path
+                                                          length:wcslen(path)*sizeof(wchar_t)
+                                                        encoding:NSUnicodeStringEncoding];
+        NSTraceInfo(TAG, @"Set %@ failed with status=%x\n", pathNSString, status);
+        [pathNSString release];
+    }
+}
+
+void setVariantFromDictionary(CFDictionaryRef dictionary,
+                              CFStringRef key,
+                              VARTYPE propertyType,
+                              LPCWSTR path,
+                              IWICMetadataQueryWriter* propertyWriter) {
+    if (dictionary && CFDictionaryContainsKey(dictionary, key)) {
+        PROPVARIANT propertyToWrite;
+        PropVariantInit(&propertyToWrite);
+        propertyToWrite.vt = propertyType;
+
+        if (propertyType == VT_UI1) {
+            propertyToWrite.bVal = (unsigned char)[(id)CFDictionaryGetValue(dictionary, key) unsignedCharValue];
+        } else if (propertyType == VT_UI2) {
+            propertyToWrite.uiVal = (unsigned short)[(id)CFDictionaryGetValue(dictionary, key) unsignedShortValue];
+        } else if (propertyType == VT_UI4) {
+            propertyToWrite.ulVal = (unsigned long)[(id)CFDictionaryGetValue(dictionary, key) unsignedLongValue];      
+        } else if (propertyType == VT_UI8) {
+            double dictionaryValue = [(id)CFDictionaryGetValue(dictionary, key) doubleValue];
+            setHighLowParts((ULARGE_INTEGER*)&propertyToWrite.uhVal, dictionaryValue);
```

Comment:
nit: it doesn't seem like the cast here is necessary (nor below where you're casting to LARGE_INTEGER*).  If the cast is necessary, maybe  setHighLowParts<ULARGE_INTEGER>(&propertyToWrite.uhVal, dictionaryValue), etc. would be cleaner?